monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/CombineLatestListObservable.scala

```diff
@@ -0,0 +1,164 @@
+    val observables: Array[Observable[(A, Int)]] = new Array(numberOfObservables)
+    observables.indices.foreach { i =>
+      observables(i) = obss(i).map(x => (x, i))
+    }
+    // MUST BE synchronized by `lock`
+    var lastAck = Continue: Future[Ack]
+    // MUST BE synchronized by `lock`
+    val elems: mutable.ArraySeq[A] = new mutable.ArraySeq(numberOfObservables)
+    // MUST BE synchronized by `lock`
+    val hasElems: Array[Boolean] = new Array(numberOfObservables)
+    hasElems.indices.foreach { i =>
+      hasElems(i) = false
+    }
+    // MUST BE synchronized by `lock`
+    var completedCount = 0
+
+    // MUST BE synchronized by `lock`
+    def rawOnNext(as: Seq[A]): Future[Ack] =
+      if (isDone) Stop
+      else {
+        var streamError = true
+        try {
+          val c = f(as)
+          streamError = false
+          out.onNext(c)
+        } catch {
+          case NonFatal(ex) if streamError =>
+            isDone = true
+            out.onError(ex)
+            Stop
+        }
+      }
+
+    // MUST BE synchronized by `lock`
+    def signalOnNext(as: Seq[A]): Future[Ack] = {
+      lastAck = lastAck match {
+        case Continue => rawOnNext(as)
+        case Stop => Stop
+        case async =>
+          async.flatMap {
+            // async execution, we have to re-sync
+            case Continue => lock.synchronized(rawOnNext(as))
+            case Stop => Stop
+          }
+      }
+
+      lastAck
+    }
+
+    def signalOnError(ex: Throwable): Unit = lock.synchronized {
+      if (!isDone) {
+        isDone = true
+        out.onError(ex)
+        lastAck = Stop
+      }
+    }
+
+    def signalOnComplete(): Unit = lock.synchronized {
+      completedCount += 1
+
+      if (completedCount == numberOfObservables && !isDone) {
+        lastAck match {
+          case Continue =>
+            isDone = true
+            out.onComplete()
+          case Stop =>
+            () // do nothing
+          case async =>
+            async.onComplete {
+              case Success(Continue) =>
+                lock.synchronized {
+                  if (!isDone) {
+                    isDone = true
+                    out.onComplete()
+                  }
+                }
+              case _ =>
+                () // do nothing
+            }
+        }
+
+        lastAck = Stop
+      }
+    }
+
+    val composite = CompositeCancelable()
+
+    observables.foreach { obs =>
+      composite += obs.unsafeSubscribeFn(new Subscriber[(A, Int)] {
+        implicit val scheduler: Scheduler = out.scheduler
+
+        def onNext(elemAndIndex: (A, Int)): Future[Ack] = lock.synchronized {
+          val elem: A = elemAndIndex._1
+          val index: Int = elemAndIndex._2
+          if (isDone) {
+            Stop
+          } else {
+            elems(index) = elem
+            if (!hasElems(index)) {
+              hasElems(index) = true
+            }
+
+            if (hasElems.forall(identity)) {
```

Comment:
We could probably set a flag the first time it is true so we don't have to check full array each time

Reply:
Or we store the number of `true`s in an additional val, so we never have to check the full array.